Denoising Diffusion Policy Optimization

metric-space opened the pull request:
Attempts to close #505 

Design notes:
1. To ensure that the current limitation of the whole setup to just the `DDIMScheduler` is not limiting any future person, this interface has been defined
```python
@dataclass
class DDPOPipelineOutput(object):
    images: torch.Tensor
    latents: torch.Tensor
    log_probs: torch.Tensor


@dataclass
class DDPOSchedulerOutput(object):
    latents: torch.Tensor
    log_probs: torch.Tensor


class DDPOStableDiffusionPipeline(StableDiffusionPipeline):
    def __call__(self, *args, **kwargs) -> DDPOPipelineOutput:
        raise NotImplementedError

    def scheduler_step(self, *args, **kwargs) -> DDPOSchedulerOutput:
        raise NotImplementedError
```
default/example implementation of this implementation which is straight out of the original code has been made available

2. It is on the user to provide a reward function and a prompt function. 
3. In addition, a function to deal with logging out the sampled images and rewards is now on the user as well

sayakpaul commented:
Hey @metric-space. Thanks so much for starting to work on this! 

I am Sayak, one of the maintainers of [diffusers](https://github.com/huggingface/diffusers). Please let me know when we can review your PR. We'd be more than happy to include a mention of your implementation from our official docs :)

metric-space commented:
Update on this: as opposed to my first message on how everything is broken, this works now (atleast locally). There's a lot more code to breakdown and section off to make things cleaner and broadly applicable

metric-space commented:
Hey folks, I think I am ready to get this reviewed. I am not sure why `github-actions` isn't pulling `diffusers` as a dependency, I've inferred that this is a github-action step configurable by the maintainers. But I could be wrong
Apologies it took this long to open this up for a review
As a courtesy, pinging @sayakpaul and @younesbelkada

sayakpaul reviewed (commented):
Design-wise, this is blowing my mind away!

Really hats off!

Could we maybe sneak peek into some results to better understand the implementation correctness? 👁️

metric-space commented:
@sayakpaul thank you and that's a fair ask.

Offlate my access  to a nifty A100 GPU machine is sparse and for this config 

key config params: train and sample batch is 3

```
    seed: 42
    logdir: logs
    log_with: wandb
    project_kwargs/logging_dir: ./logs
    project_kwargs/automatic_checkpoint_naming: true
    project_kwargs/total_limit: 5
    project_kwargs/project_dir: ./save
    tracker_project_name: stable_diffusion_training
    num_epochs: 200
    save_freq: 1
    num_checkpoint_limit: 5
    mixed_precision: fp16
    allow_tf32: true
    resume_from: ''
    use_lora: true
    pretrained_model: runwayml/stable-diffusion-v1-5
    pretrained_revision: main
    sample_num_steps: 30
    sample_eta: 1.0
    sample_guidance_scale: 5.0
    sample_batch_size: 3
    sample_num_batches_per_epoch: 2
    train_batch_size: 3
    train_use_8bit_adam: false
    train_learning_rate: 0.0003
    train_adam_beta1: 0.9
    train_adam_beta2: 0.999
    train_adam_weight_decay: 0.0001
    train_adam_epsilon: 1.0e-08
    train_gradient_accumulation_steps: 1
    train_max_grad_norm: 1.0
    train_num_inner_epochs: 1
    train_cfg: true
    train_adv_clip_max: 5
    train_clip_range: 0.0001
    train_timestep_fraction: 1.0
    per_prompt_stat_tracking: false
    per_prompt_stat_tracking_buffer_size: 16
    per_prompt_stat_tracking_min_count: 16
    async_reward_computation: false
    negative_prompts: 
   ```
   
![rewards](https://github.com/lvwerra/trl/assets/6382230/17fc5e11-e0ee-4cb3-acbb-78a6dd5b5622)
![loss](https://github.com/lvwerra/trl/assets/6382230/00cba922-a599-4562-b2de-fa2a101680e7)

sayakpaul commented:
@younesbelkada possible for you to chug a training run on an A100?

metric-space commented:
I've managed to get another round with an A100 on my side. I have updated the example script with the config params I have running  right now. Will update this thread with the run results after it's done which should be in about an hour from now I think

metric-space commented:
Update to my last message: (still running ~70 epochs in out of 200) so the current params are:

```py

    config = DDPOConfig(
        num_epochs=200,
        train_gradient_accumulation_steps=1,
        sample_num_steps=50,
        sample_batch_size=6,
        train_batch_size=3,
        sample_num_batches_per_epoch=4,
        per_prompt_stat_tracking=True,
        per_prompt_stat_tracking_buffer_size=32,
       ...)
```

![rewards_2](https://github.com/lvwerra/trl/assets/6382230/26a9283a-f072-4ee5-97e2-5f695d238dab)
![epoch_2](https://github.com/lvwerra/trl/assets/6382230/5343ef3b-8917-4564-a9be-795c7ecfcf23)
![loss_2](https://github.com/lvwerra/trl/assets/6382230/77dcc34c-4ce7-4f35-8268-e0c5f6d51d48)
![reward_std_2](https://github.com/lvwerra/trl/assets/6382230/83eedb44-c5dd-42df-8756-e5fab2abac6d)

sayakpaul commented:
Possible to share a couple of visual examples?

younesbelkada commented:
The reward is growing up, very nice! Do you have any example of generated images out of curiosity? 🙏

metric-space commented:
![zebra_6_82](https://github.com/lvwerra/trl/assets/6382230/92e1e4af-2d5f-40af-815d-c65d694b7a78)
![cow_6_55](https://github.com/lvwerra/trl/assets/6382230/65ff7220-9477-4b41-a990-c6ad6ba940fb)
![bird_6_72](https://github.com/lvwerra/trl/assets/6382230/9622a8f9-a9a4-4023-b3f6-4694cf147715)
![bee_6_04](https://github.com/lvwerra/trl/assets/6382230/4489d700-0890-4126-baee-cc6e13e8be23)
![horse_1](https://github.com/lvwerra/trl/assets/6382230/20b76235-7929-4761-b8e2-d7712b7d16e5)

younesbelkada commented on docs/source/ddpo_trainer.mdx:
```diff
@@ -0,0 +1,61 @@
+
+# Getting started with Stable Diffusion finetuning with reinforcement learning
+
+The machinery for finetuning of Stable Diffusion models with reinforcement learning makes heavy use of HuggingFace's `diffuser`
```
```suggestion
The machinery for finetuning of Stable Diffusion models with reinforcement learning makes heavy use of HuggingFace's `diffusers`
```

younesbelkada commented on docs/source/ddpo_trainer.mdx:
```diff
@@ -0,0 +1,61 @@
+# Getting started with Stable Diffusion finetuning with reinforcement learning
+
+The machinery for finetuning of Stable Diffusion models with reinforcement learning makes heavy use of HuggingFace's `diffuser`
+library. A reason for  stating this is that getting started requires a bit of familiarity with the `diffuser` library concepts, mainly two of them.
```
```suggestion
library. A reason for  stating this is that getting started requires a bit of familiarity with the `diffusers` library concepts, mainly two of them.
```

sayakpaul commented:
I didn't catch the wandb piece. I think you're logging a couple of validation images on WandB as well, right?

younesbelkada commented on examples/scripts/stable_diffusion_tuning.py:
```diff
@@ -0,0 +1,160 @@
+
+
+class AestheticScorer(torch.nn.Module):
+    def __init__(self, dtype, cache=".", weights_fname="sac+logos+ava1-l14-linearMSE.pth"):
+        super().__init__()
+        self.clip = CLIPModel.from_pretrained("openai/clip-vit-large-patch14")
+        self.processor = CLIPProcessor.from_pretrained("openai/clip-vit-large-patch14")
+        self.mlp = MLP()
+        self.loadpath = os.path.join(cache, weights_fname)
+        if not os.path.exists(self.loadpath):
+            url = (
+                "https://github.com/christophschuhmann/"
+                f"improved-aesthetic-predictor/blob/main/{weights_fname}?raw=true"
+            )
+            r = requests.get(url)
+
+            with open(self.loadpath, "wb") as f:
+                f.write(r.content)
+
+        state_dict = torch.load(self.loadpath)
```
I have a better suggestion for the name. You can store them on a HF model repo on the hub (i.e. upload the weights on a new fresh repo under your namespace) and use the `hf_hub_download` method from `huggingface_hub` library. Let us assume you have stored these weights under `myusername/ddpo-aestetic-scorer` repository, this block would look like:


```suggestion
    def __init__(self, dtype, hub_model_id="myusername/ddpo-aestetic-scorer"):
        super().__init__()
        self.clip = CLIPModel.from_pretrained("openai/clip-vit-large-patch14")
        self.processor = CLIPProcessor.from_pretrained("openai/clip-vit-large-patch14")
        self.mlp = MLP()
        cached_path = hf_hub_download(hub_model_id, "sac+logos+ava1-l14-linearMSE.pth")
        state_dict = torch.load(cached_path)
```

`hf_hub_download` method will directly save the downloaded file on the cache and return the absolute path of the cached file

younesbelkada commented on requirements.txt:
```diff
@@ -4,3 +4,5 @@ tqdm
 transformers
 accelerate
 peft>=0.3.0
+diffusers>=0.18.0
```
Actually I think what is better to do is to put diffusers here: https://github.com/lvwerra/trl/blob/main/setup.py#L70 in the extra test dependency and make the dependency with diffusers optional, meaning we want to protect all diffusers related import with a new method `is_diffusers_available()`, similarly as peft: https://github.com/lvwerra/trl/blob/main/trl/import_utils.py#L24

younesbelkada commented on tests/test_ddpo_trainer.py:
```diff
@@ -0,0 +1,77 @@
+import gc
```
(nit) can you add the copy right here with your GH handle + Hugging Face team: https://github.com/lvwerra/trl/blob/main/trl/import_utils.py#L1

younesbelkada commented on examples/scripts/stable_diffusion_tuning.py:
```diff
@@ -0,0 +1,160 @@
+import os
```
(nit) can you add the copy right here with your GH handle + Hugging Face team: https://github.com/lvwerra/trl/blob/main/trl/import_utils.py#L1

younesbelkada commented on trl/models/modelling_sd_base.py:
```diff
@@ -0,0 +1,405 @@
+
+import numpy as np
+import torch
+from diffusers import DDIMScheduler, StableDiffusionPipeline
+from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion import rescale_noise_cfg
```
these imports needs to be protected with `is_diffusers_available()` similarly as PEFT: https://github.com/lvwerra/trl/blob/main/trl/import_utils.py#L24

metric-space replied:
I have another question for you: 

I am not sure if this is the `below` comment you mentioned here

> After thinking a bit, we should maybe protect that whole import with is_diffusers_available() instead of my comment below, what do you think?

If not, my question: 
since the `is_diffusers_available` is responsible for conditionally importing all the functions related to DDPO at the __init__ level (at the main `__init__`, the trainer directory level `__init__` and models directory level `__init__`), would guarding it here be an overkill? I guess I'm not able to reason why we would want it here, if you could, please educate me on the use-case

sayakpaul replied:
We do it this way to keep the total dependency of the library lightweight. So, if a user doesn't need diffusion related utilities, they don't have to `diffusers`.

metric-space replied:
> We do it this way to keep the total dependency of the library lightweight
I understand this portion. I guess I'm asking if it's the guard is necessary at this file level if I've already put in guard at the __init__ levels?

younesbelkada commented on trl/models/modelling_sd_base.py:
```diff
@@ -0,0 +1,405 @@
+
+@dataclass
+class DDPOPipelineOutput(object):
+    images: torch.Tensor
```
can you please add some docstrings here 🙏

younesbelkada commented on trl/models/modelling_sd_base.py:
```diff
@@ -0,0 +1,405 @@
+
+@dataclass
+class DDPOSchedulerOutput(object):
+    latents: torch.Tensor
```
same comment as above

younesbelkada commented on trl/models/__init__.py:
```diff
@@ -15,6 +15,13 @@
 # limitations under the License.
 from .modeling_base import PreTrainedModelWrapper, create_reference_model
 from .modeling_value_head import AutoModelForCausalLMWithValueHead, AutoModelForSeq2SeqLMWithValueHead
+from .modelling_sd_base import (
```
After thinking a bit, we should maybe protect that whole import with `is_diffusers_available()` instead of my comment below, what do you think?

sayakpaul replied:
Yes, that is the way to go here.

younesbelkada commented on trl/trainer/__init__.py:
```diff
@@ -21,6 +21,8 @@
 # isort: on
 
 from .base import BaseTrainer
+from .ddpo_config import DDPOConfig
+from .ddpo_trainer import DDPOTrainer
```
import to protect with `is_diffusers_available()`

younesbelkada commented on trl/trainer/ddpo_trainer.py:
```diff
@@ -0,0 +1,620 @@
+# NOTE: code heavily inspired by https://github.com/kvablack/ddpo-pytorch
```
Copy right to add here

sayakpaul replied:
Yeah should include the license and then include the reference.

younesbelkada replied:
I would just add something like:

```python
# Copyright 2023 DDPO-pytorch authors, metric-space,  The HuggingFace Team. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
```

the DDPO-pytorch repo is under MIT license so should be all good

metric-space commented:
@sayakpaul  The images I logged on here are from the samples being generated pre-training step, while one has the power to log all of these out to wandb, I do the last bunch of a samples generated per step as of now for this particular stepup

Just to be clear, my reasoning is that the the generated samples whose associated rewards is being for training is the same as validation samples (in my head). Please do tell me if I'm wrong

younesbelkada reviewed (commented):
The results are astonishing! Thanks for your huge work @metric-space  ! I left few initial high level comments before deep diving into the code, I think we should make diffusers dependency optional and import the relevant diffusers-related objects by protecting the import with `is_diffusers_available()`. Please have a look at my comments and let me know what do you think, Thanks !

metric-space commented:
@younesbelkada Thank you for the review and the comments. Do expect the changes to be put in sometime tomorrow

younesbelkada commented:
Thanks very much @metric-space , please let me know if you need any help !

metric-space commented on trl/models/modelling_sd_base.py:
```diff
@@ -1,5 +1,19 @@
+# Copyright 2023 DDPO-pytorch authors (Kevin Black), The HuggingFace Team, metric-space. All rights reserved.
```
@younesbelkada just a note here: I changed the order of authors names as the default implementation code  is mostly from the diffusers library

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/trl/pr_508). All of your documentation changes will be reflected on that endpoint.

metric-space commented:
Most of the changes are in. Got hit with an accelerate error during testing which I'm guessing is because it (accelerate) is used in other tests with slightly different settings. Seems like trying to side-step the problem with mixed_precision=None is running into GPU memory limit problems. Looking into it but won't be able get back on this till tomorrow

sayakpaul commented:
I will do a more thorough review soon. Just setting a reminder for myself. 

Thanks for your hard work!

metric-space commented:
It feels like all the previous issues have addressed with the failing tests. I seem to be encountering memory issues on windows machines, is it possible to throw more cpu memory at specifically window machines?

sayakpaul commented on docs/source/ddpo_trainer.mdx:
```diff
@@ -0,0 +1,61 @@
+
+# Getting started with Stable Diffusion finetuning with reinforcement learning
+
+The machinery for finetuning of Stable Diffusion models with reinforcement learning makes heavy use of HuggingFace's `diffusers`
+library. A reason for  stating this is that getting started requires a bit of familiarity with the `diffusers` library concepts, mainly two of them.
+`Pipeline` and `Scheduler`. 
```
```suggestion

The machinery for finetuning of Stable Diffusion models with reinforcement learning makes heavy use of HuggingFace's `diffusers`
library. A reason for  stating this is that getting started requires a bit of familiarity with the `diffusers` library concepts, mainly two of them - pipelines and schedulers.
```

sayakpaul commented on docs/source/ddpo_trainer.mdx:
```diff
@@ -0,0 +1,61 @@
+`Pipeline` and `Scheduler`. 
+Right out of the box (`diffusers` library), there isn't a `Pipeline` nor a `Scheduler` instance that is suitable for finetuning with reinforcement learning. Some adjustments need to made. 
+
+There is an interface that is provided by this library that are required to be implemented in order to be use the `DDPOTrainer`, which is the main machinery for finetuning. 
+There is a default implementation of this interface that you can use out of the box. Assuming the default implementation is sufficient and/or to get thing moving, refer to the training example alongside this readme. 
```
```suggestion
There is an interface that is provided by this library that is required to be implemented to be used with the `DDPOTrainer`, which is the main machinery for fine-tuning Stable Diffusion with reinforcement learning. 
There is a default implementation of this interface that you can use out of the box. Assuming the default implementation is sufficient and/or to get things moving, refer to the training example alongside this guide. 
```

sayakpaul commented on docs/source/ddpo_trainer.mdx:
```diff
@@ -0,0 +1,61 @@
+While there is an explicit class interface for the pipeline, the scheduler is constrained by a method signature associated with the pipeline class. The trainer will attempt to get the scheduler
+via this pipeline associated method rather than directly. This was the most minimal way forward and the only constraint is that a particular required output type.
+
+For a more detailed look into the interface and the default scheduler and pipeline implementation, go [here](https://github.com/lvwerra/trl/models/modelling_sd_base.py)
```
I think the absolute URL needs to be corrected here.

sayakpaul commented on docs/source/ddpo_trainer.mdx:
```diff
@@ -0,0 +1,61 @@
+
+For a more detailed look into the interface and the default scheduler and pipeline implementation, go [here](https://github.com/lvwerra/trl/models/modelling_sd_base.py)
+
+Also in addition, there is the expectation of providing a reward function and a prompt function. The reward function is used to evaluate the generated images  and the prompt function is used to generate the prompts that are used to generate the images.
```
If we're fine-tuning on a dataset of image-caption pairs, then why do we need the prompt generation function?

metric-space replied:
I may have not understood the question properly but while finetuning on a dataset of image-caption pairs is possible though indirectly, we're counting on the reward function to be the main force behind the fine-tuning.  It feels like I'm preaching to the choir, so not sure what I'm missing here

sayakpaul commented on docs/source/ddpo_trainer.mdx:
```diff
@@ -0,0 +1,61 @@
+
+Also in addition, there is the expectation of providing a reward function and a prompt function. The reward function is used to evaluate the generated images  and the prompt function is used to generate the prompts that are used to generate the images.
+
+# Setting up the image logging hook function
```
Since this is still the beginning of the document, I would first show the readers a minimal run down of the commands needed to use the trainer. Then I would proceed to this.

sayakpaul commented on docs/source/ddpo_trainer.mdx:
````diff
@@ -0,0 +1,61 @@
+
+# Setting up the image logging hook function
+
+Expect the function to be given a list of lists of the form
+```python
+[[image, prompt, prompt_metadata, rewards], ...]
+
+```
+and `image`, `prompt`, `prompt_metadata`, `rewards` are batched.
+The last list in the lists of lists represents the last sample batch. You are likely to want to log this one
+While you are free to log however you want the use of `wandb` or `tensorboard` is recommended.
+
+Example code for logging sampled images with `wandb` is given below.
````
Worth explaining what `prompt_metadata` and `rewards` are.

sayakpaul commented on docs/source/ddpo_trainer.mdx:
```diff
@@ -0,0 +1,61 @@
+- The configurable sample batch size must be divisible by the configurable train batch size
+- The configurable sample batch size must be divisible by both the configurable gradient accumulation steps and the configurable accelerator processes count
+
+For a working example please refer to `stable_diffusion_tuning.py` in the `examples/scripts` directory.
```
Shouldn't we also include links that show the reader how to use the `stable_diffusion_tuning.py` script to get a run started?

sayakpaul commented on examples/scripts/stable_diffusion_tuning.py:
```diff
@@ -0,0 +1,165 @@
+    return _fn
+
+
+animals = [
```
Could add a comment here.

sayakpaul commented on examples/scripts/stable_diffusion_tuning.py:
```diff
@@ -0,0 +1,165 @@
+
+
+def image_outputs_logger(image_data, global_step, accelerate_logger):
+    # extract the last one
```
The comment is not very clear.

sayakpaul commented on examples/scripts/stable_diffusion_tuning.py:
```diff
@@ -0,0 +1,165 @@
+
+
+if __name__ == "__main__":
+    config = DDPOConfig(
+        num_epochs=200,
+        train_gradient_accumulation_steps=1,
+        sample_num_steps=50,
+        sample_batch_size=6,
+        train_batch_size=3,
+        sample_num_batches_per_epoch=4,
+        per_prompt_stat_tracking=True,
+        per_prompt_stat_tracking_buffer_size=32,
+        tracker_project_name="stable_diffusion_training",
+        log_with="wandb",
+        project_kwargs={
+            "logging_dir": "./logs",
+            "automatic_checkpoint_naming": True,
+            "total_limit": 5,
+            "project_dir": "./save",
+        },
+    )
+    pretrained_model = "runwayml/stable-diffusion-v1-5"
+    pretrained_revision = "main"
```
Could benefit from CLI arg parsing.

sayakpaul commented on trl/models/modelling_sd_base.py:
```diff
@@ -0,0 +1,441 @@
+    return t.reshape(t.shape + (1,) * (len(shape) - t.ndim)).broadcast_to(shape)
+
+
+class DefaultDDPOScheduler(DDIMScheduler):
```
Could add a comment here to make the reader aware of the reason why this subclassing is required.

sayakpaul commented on trl/models/modelling_sd_base.py:
```diff
@@ -0,0 +1,441 @@
+        guidance_rescale: float = 0.0,
+    ):
+        r"""
+        Function invoked when calling the pipeline for generation.  Args: prompt (`str` or `List[str]`, *optional*): The prompt or prompts to guide the image generation. If not defined, one has to pass `prompt_embeds`.  instead.  height (`int`, *optional*, defaults to self.unet.config.sample_size * self.vae_scale_factor): The height in pixels of the generated image.
```
Same. Worth adding a line or two to explain the differences we're introducing in this class w.r.t the standard pipeline.

sayakpaul reviewed (commented):
The `diffusers` related bits look quite nice to me. 

From the users' point of view, I felt a lack of documentation showing me how to use the trainer on a dataset, etc. So, some guidance explaining that would be much appreciated. 

I will let the `trl` folks comment on the other bits.

metric-space commented:
@sayakpaul Thank you for the review and the comments. Will work on the corrections shortly

metric-space commented:
As promised, I believe all the requested edits/changes are in. Ball's in your court folks

lvwerra commented on trl/trainer/ddpo_trainer.py:
```diff
@@ -0,0 +1,633 @@
+        self.trainable_layers, self.optimizer = self.accelerator.prepare(trainable_layers, self.optimizer)
+
+        if self.config.async_reward_computation:
+            self.executor = futures.ThreadPoolExecutor(max_workers=2)
```
should the `max_workers` be configurable?

lvwerra commented on trl/trainer/ddpo_trainer.py:
```diff
@@ -0,0 +1,633 @@
+        del samples["prompt_ids"]
+
+        total_batch_size, num_timesteps = samples["timesteps"].shape
+        assert total_batch_size == self.config.sample_batch_size * self.config.sample_num_batches_per_epoch
+        assert num_timesteps == self.config.sample_num_steps
```
Can you make those proper `ValueErrors` (or an other appropriate error class) with helpful error messages?

metric-space replied:
Removed all asserts except  one that I reason should be enough to catch all in case something's off. Also  everything should follow from calculated settings

lvwerra commented on trl/trainer/ddpo_trainer.py:
```diff
@@ -0,0 +1,633 @@
+            self.sd_pipeline.unet.train()
+            global_step = self._train_batched_samples(inner_epoch, epoch, global_step, samples_batched)
+            # ensure optimization step at the end of the inner epoch
+            assert self.accelerator.sync_gradients
```
same here, use errors instead of asserts.

lvwerra commented on trl/trainer/ddpo_trainer.py:
```diff
@@ -0,0 +1,633 @@
+
+    def _setup_optimizer(self, trainable_layers_parameters):
+        if self.config.train_use_8bit_adam:
+            import bitsandbytes
```
since `bitsandbytes` optional can we replace this with an try/except and raise an `ImportError` telling the user to install the library if they want to use 8bit-adam?

metric-space replied:
I almost forgot about this, I think the `__post_init__` of the DDPO config deals with this

lvwerra commented on trl/trainer/ddpo_trainer.py:
```diff
@@ -0,0 +1,633 @@
+
+                # Checks if the accelerator has performed an optimization step behind the scenes
+                if self.accelerator.sync_gradients:
+                    assert (j == self.num_train_timesteps - 1) and (
```
again, please no `asssert` :)

lvwerra commented on docs/source/ddpo_trainer.mdx:
```diff
@@ -0,0 +1,103 @@
+# Credits
```
Can you only add the title with a top level section (`#`) with the title being `Denoising Diffusion Policy Optimization` and then adjusting the subsections accordingly? I have a slight preference to put credits/references at the end of the document but no strong opinion.

lvwerra reviewed (commented):
Hi @metric-space, the PR is in a really good shape. I left some feedback which are mostly small nits. The main discussion is around how agnostic/flexible some choices are. I think we can merge this soon, great job!

metric-space commented:
@lvwerra  Thank you for the review and the feedback. Will try to address all of these by this coming Monday

metric-space commented on examples/scripts/stable_diffusion_tuning.py:
```diff
@@ -0,0 +1,211 @@
+        project_kwargs=project_kwargs,
+    )
+
+    pretrained_model = args.pretrained_model
+    pretrained_revision = args.pretrained_revision
+
+    pipeline = DefaultDDPOPipeline.from_pretrained(pretrained_model, revision=pretrained_revision)
```
Not a bit. Very redundant, this one really flew by me

metric-space commented on trl/trainer/ddpo_trainer.py:
```diff
@@ -0,0 +1,633 @@
+        if not config.run_name:
+            config.run_name = uuid.uuid4().hex
+        else:
+            config.run_name += "_" + uuid.uuid4().hex
```
I did not think this through, correcting this

metric-space commented on trl/trainer/ddpo_trainer.py:
```diff
@@ -0,0 +1,633 @@
+        if config.use_lora:
+            self.sd_pipeline.unet.to(self.accelerator.device, dtype=inference_dtype)
+
+        if config.use_lora:
+            # Set correct lora layers
+            lora_attn_procs = {}
+            for name in self.sd_pipeline.unet.attn_processors.keys():
+                cross_attention_dim = (
+                    None if name.endswith("attn1.processor") else self.sd_pipeline.unet.config.cross_attention_dim
+                )
+                if name.startswith("mid_block"):
+                    hidden_size = self.sd_pipeline.unet.config.block_out_channels[-1]
+                elif name.startswith("up_blocks"):
+                    block_id = int(name[len("up_blocks.")])
+                    hidden_size = list(reversed(self.sd_pipeline.unet.config.block_out_channels))[block_id]
+                elif name.startswith("down_blocks"):
+                    block_id = int(name[len("down_blocks.")])
+                    hidden_size = self.sd_pipeline.unet.config.block_out_channels[block_id]
+
+                lora_attn_procs[name] = LoRAAttnProcessor(
+                    hidden_size=hidden_size, cross_attention_dim=cross_attention_dim
+                )
+            self.sd_pipeline.unet.set_attn_processor(lora_attn_procs)
+            trainable_layers = AttnProcsLayers(self.sd_pipeline.unet.attn_processors)
```
I think because all of this is specific to stable diffusion pipelines and all of them use the 2d unet, I believe this should work with all of them

metric-space commented on trl/trainer/ddpo_trainer.py:
```diff
@@ -0,0 +1,633 @@
+        self.sd_pipeline = sd_pipeline
+
+        # freeze parameters of models to save more memory
+        self.sd_pipeline.vae.requires_grad_(False)
+        self.sd_pipeline.text_encoder.requires_grad_(False)
+        self.sd_pipeline.unet.requires_grad_(not config.use_lora)
```
I believe these should be frozen universally. At this point cannot imagine a scenario where the user would want otherwise. That said, this wouldn't be the first time lacked imagination and foresight

metric-space commented on trl/trainer/ddpo_trainer.py:
```diff
@@ -0,0 +1,633 @@
+        self.sd_pipeline.vae.to(self.accelerator.device, dtype=inference_dtype)
+        self.sd_pipeline.text_encoder.to(self.accelerator.device, dtype=inference_dtype)
+        if config.use_lora:
+            self.sd_pipeline.unet.to(self.accelerator.device, dtype=inference_dtype)
```
I believe this is correct, the lack of triggers scaling issues like such `ValueError: Attempting to unscale FP16 gradients`. Will be relocating this line

metric-space replied:
With the latest results, will be putting this on the accelerator device for non-lora as well

metric-space commented:
@lvwerra Almost all of your feedback has been addressed. I'm stuck on trying to figure out why `non-lora` is not converging at all despite ad-hoc corrections from device type to precision type. Atm addressing this. I won't be able to deliver work on this till this Friday 

As a side, I have refactored the main interface to be implemented by the user to expand the use of the pipeline type. The new interface is not as pretty as it's former self, so was hoping to get your take on it
While it may not be pretty, it's utility is 5 fold that of it's past self

metric-space commented:
Update: from continued failed training, it seems an unlikely yet strong culprit has emerged as the main reason as to why non-lora training is providing horrendous results. Because training surface is huge, if the reward scorer is inept preventing the complete screwing up of the unet is almost impossible.
With the current aesthetic scorer, there are quite a lot of feedback it gives that seems quite generous when it should not. Testing hypothesis

metric-space commented:
Update: reward function problem  Hypothesis proven False. Looking at this as a hyperparameter tuning problem

metric-space commented:
Update: it was a hyper-parameter tuning problem, for future reference, the config that got things moving when it comes to `non-lora` territory (the main portions)

```python
use_lora=False,
mixed_precision=None,
train_clip_range=5e-5,
train_learning_rate=0.00005
```

and sample_batch_size and train_batch_size are the same, 4

lvwerra commented:
Very interesting! Does that mean the PR is ready for another review?

metric-space commented:
@lvwerra Yes and please go ahead

lvwerra commented on docs/source/ddpo_trainer.mdx:
````diff
@@ -0,0 +1,104 @@
+    )
+
+```
+
````
We had some requests recently to also show how to use the model after training. Could you add a small snippet that given the user pushed the model to the hub how it can be loaded and an image generated?

lvwerra replied:
I think in the example you could link your trained model!

lvwerra reviewed (approved):
Hi this looks all pretty good to me. Left a few small nits and requests:
- Can you add the classes also to the section `trainer`/`Trainer Classes` where the api as documented?
- The implementation currently only supports SD - I think that's fine but we need to be very upfront about it. So I added a few comments to make it clearer.
- Add a little example of how to use the trained model.

@younesbelkada it would be great if you could also have a quick look and @sayakpaul if you could confirm that this fits well into the `diffusers` ecosystem.

I already add an approval assuming you'll address the above points. Awesome job! 🚀 

_Note:_ you'll need to merge main into your branch to fix the merge conflicts.

sayakpaul commented:
> @younesbelkada it would be great if you could also have a quick look and @sayakpaul if you could confirm that this fits well into the diffusers ecosystem.

@lvwerra yes, I can confirm that it utilizes `diffusers` very nicely. I can confirm this quickly because I had already reviewed the `diffusers` specific bits earlier.

I think it would be very nice to add a blog documenting all the crucial findings and potentially incorporate LoRA into a the mix to make it more accessible.

younesbelkada commented on trl/models/modeling_sd_base.py:
```diff
@@ -0,0 +1,517 @@
+
+
+def _left_broadcast(t, shape):
+    assert t.ndim <= len(shape)
```
```suggestion
    if t.ndim > len(shape):
        raise ValueError("`ndim` should be less or equal than `shape`") 
```
Can you also elaborate what are t and shape?

younesbelkada replied:
just one line would be sufficient IMO

younesbelkada commented on trl/models/modeling_sd_base.py:
```diff
@@ -0,0 +1,517 @@
+            self.alphas_cumprod.gather(0, prev_timestep.cpu()),
+            self.final_alpha_cumprod,
+        )
+        alpha_prod_t = _left_broadcast(alpha_prod_t, sample.shape).to(sample.device)
```
Ah I understand now the purpose of it!

younesbelkada reviewed (commented):
Thanks for your great work! I left few tiny comments, please have a look a them before merging, really awesome work @metric-space !

younesbelkada commented:
Also we should maybe change key words in the main README a bit to emphasize that TRL supports also diffusion models instead of language models only, e.g. something like "Full stack transformer language models and diffusion models with reinforcement learning." instead of "Full stack transformer language models with reinforcement learning.". What do you think?

metric-space commented:
Hey folks, thank you for the comments, the reviews and the feedback and most of all your patience. Will be putting the changes over the weekend, so expect the changes all in by Monday

sayakpaul commented:
@metric-space the pleasure is ours! 

I would very much like to welcome you to author a blog about it on hf.co/blog so that community is aware of this and get a handle of what you have tried, what has worked and what hasn't. WDYT? 

Happy to set up a Slack channel with you for better coordination :)

metric-space commented:
@sayakpaul :ultra-thumbs-up: I welcome the opportunity and would be happy to do write up. The slack channel stuff sounds good to me, I just enabled my email to be displayed on my public profile so should be there till I get hold of the invite link for the slack channel

sayakpaul commented:
Thanks much! Watch out for an invite from one of our admins!

kashif commented on trl/trainer/ddpo_trainer.py:
```diff
@@ -0,0 +1,620 @@
+            dynamic_ncols=True,
+        )
+
+        # For mixed precision training we cast all non-trainable weigths (vae, non-lora text_encoder and non-lora unet) to half-precision
```
```suggestion
        # For mixed precision training we cast all non-trainable weights (vae, non-lora text_encoder and non-lora unet) to half-precision
```

metric-space commented on examples/scripts/stable_diffusion_tuning.py:
```diff
@@ -0,0 +1,206 @@
+        self.clip = CLIPModel.from_pretrained("openai/clip-vit-large-patch14")
+        self.processor = CLIPProcessor.from_pretrained("openai/clip-vit-large-patch14")
+        self.mlp = MLP()
+        cached_path = hf_hub_download(hub_model_id, "sac+logos+ava1-l14-linearMSE.pth")
```
My take is a bit different in terms of where the default is place. Please take a look and let me know what you think of it

metric-space commented:
Hey there folks,

I've addressed all the feedback but also have made a refactor  (my apologies) that as a consequence requires your attention again. Now the interface and it's implementation really does minimize how much the trainer knows about the internals. I believe this is what @lvwerra  was getting at via the questions asked

The review will lengthen how much more longer it will take this merge this into main, while I'm aware of this but reasoning of the code became a little too cumbersome and whatever hangups I had abut shifting model-specific code into the interface implementation melted away while working on `saving/loading`. Also I grew ashamed of the code

In addition the pipeline and scheduler that I stuck to putting in class form is now back to function form as saving/loading became a headache and thought this was the fastest way forward

I do need some help in sticking the pictures for before and after added to your CDN, currently they are not rendering on the docs

metric-space commented:
Before I forget, I ran a new training session with the new refactored code and uploaded the weights to https://huggingface.co/metric-space/ddpo-finetuned-sd-model and the `after` images are from this model

![W B Chart 2023-08-21, 4 07 30 a m](https://github.com/huggingface/trl/assets/6382230/3da2d7e8-6b6a-4957-8df7-3237ba204fba)

metric-space commented:
Noticed a blip, corrected it, should be good for the review

younesbelkada reviewed (approved):
Looking great ! Only have a single comment, thanks for your great work @metric-space !

younesbelkada commented on docs/source/ddpo_trainer.mdx:
```diff
@@ -0,0 +1,141 @@
+
+| Before | After DDPO finetuning |
+| --- | --- |
+| ![Pre squirrel](/resources/images/pre_squirrel.png) | ![Post squirrel](/resources/images/post_squirrel.png) |
+| ![Pre crab](/resources/images/pre_crab.png) | ![Post crab](/resources/images/post_crab.png) |
+| ![Pre starfish](/resources/images/pre_starfish.png) | ![Post starfish](/resources/images/post_starfish.png) |
```
I pushed the images here: https://huggingface.co/datasets/trl-internal-testing/example-images/tree/main/images can you update the links with the ones on the Hub, similarly as https://github.com/huggingface/trl/blob/main/docs/source/detoxifying_a_lm.mdx#L86 for example

younesbelkada replied:
I kept the same filenames: so replacing `/resources/images/xxxx` by `https://huggingface.co/datasets/trl-internal-testing/example-images/resolve/main/images/xxxx` should do the trick
<img width="898" alt="Screenshot 2023-08-21 at 11 40 25" src="https://github.com/huggingface/trl/assets/49240599/30e5a156-cde8-45de-842d-3aeaf5edfba4">

metric-space commented:
@younesbelkada thank you and thank you for all your help. The feedback has been address. All yours

younesbelkada reviewed (approved):
Big thanks again @metric-space !
